### PR TITLE
Upgrade to ZIO 2.0.0-RC5

### PR DESCRIPTION
**Pull request by gavares**
Still one test failing in LeaseLockSpec related to a fiber not being
interrupted when clock skew occurs

TODO: remove debugging code in LeaseLockSpec

### CODE REVIEW

**Comment by gavares** [comment]
After some debugging it's obvious that `f2` in the failing test isn't using `otherClock` which is supplied to it in the `provideSomeEnvironment(_ ++ otherClock)` call.

File: examples/logs-example/src/main/scala/com/coralogix/zio/k8s/examples/logs/LogsExample.scala [hunk 1]
```diff
@@ -42,15 +44,14 @@ object LogsExample extends ZIOAppDefault {
...
-      .exitCode
+    program.provideSome[ZIOAppArgs](pods)
   }
 
   private def tailLogs(
+    pods: Pods,
     podName: String,
     containerName: Option[String]
-  ): ZIO[Pods with Console, K8sFailure, Unit] =
+  ): IO[K8sFailure, Unit] =
```
Comment (vigoo) [inline examples/logs-example/src/main/scala/com/coralogix/zio/k8s/examples/logs/LogsExample.scala:54]: I would keep this example using `Pods` as a dependency passed through environment as before. The RC5 upgrade itself is not a reason for change that.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala [hunk 2]
```diff
@@ -13,7 +13,7 @@ import javax.net.ssl.{ KeyManager, KeyManagerFactory }
 
 private object KeyManagers {
 
-  private def getDefaultKeyStore: ZIO[System, Throwable, KeyStore] =
+  private def getDefaultKeyStore: ZIO[Scope, Throwable, KeyStore] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala:16]: There's no need to extend the managed scope of this outside of `getDefaultKeyStore`. The old version was `ZIO`, meaning that it handled the opening and closing of the key file inside as a black box, not leaking it through its interface. Let's keep it that way.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala [hunk 3]
```diff
@@ -22,7 +22,7 @@ private object KeyManagers {
       _                   <-
         propertyKeyStoreFile match {
           case Some(file) =>
-            ZManaged.fromAutoCloseable(ZIO.attempt(new FileInputStream(file))).use { stream =>
+            ZIO.fromAutoCloseable(ZIO.attempt(new FileInputStream(file))) flatMap { stream =>
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala:25]: The equivalent of `.use` is `ZIO.scoped { managed.flatMap { ... } }```. So you should introduce a new scope here that closes the file at the same place as the `use` ended before.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala [hunk 4]
```diff
@@ -34,12 +34,12 @@ private object KeyManagers {
     certificate: KeySource,
     key: KeySource,
     password: Option[String]
-  ): ZIO[System, Throwable, Array[KeyManager]] =
+  ): ZIO[Any with Scope, Throwable, Array[KeyManager]] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala:37]: No need to leak the scope if you do the changes I described above

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/SSL.scala [hunk 5]
```diff
@@ -1,17 +1,16 @@
...
 import java.security.cert.X509Certificate
 import javax.net.ssl.{ KeyManager, SSLContext, TrustManager, X509TrustManager }
-import zio.System
 
 private object SSL {
   def apply(
     serverCertificate: K8sServerCertificate,
     authentication: K8sAuthentication
-  ): ZIO[System with Any, Throwable, SSLContext] =
+  ): ZIO[Any with Scope, Throwable, SSLContext] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/SSL.scala:13]: Does not have to be scoped

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/SSL.scala [hunk 6]
```diff
@@ -35,8 +34,8 @@ private object SSL {
   private def secureSSLContext(
     certSource: KeySource,
     authentication: K8sAuthentication
-  ): ZIO[System with Any, Throwable, SSLContext] =
-    loadKeyStream(certSource).use { certStream =>
+  ): ZIO[Any with Scope, Throwable, SSLContext] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/SSL.scala:37]: Does not have to be scoped

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/TrustManagers.scala [hunk 7]
```diff
@@ -19,7 +19,7 @@ private object TrustManagers {
 
   private def getDefaultTrustStoreWithSecurityDir(
     secDir: Path
-  ): ZIO[Any with System, Throwable, KeyStore] =
+  ): ZIO[Any with Scope, Throwable, KeyStore] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/TrustManagers.scala:22]: Same thing about extending the scope from the original one as I described above

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/asynchttpclient/package.scala [hunk 8]
```diff
@@ -15,22 +15,17 @@ package object asynchttpclient {
...
-                                        case K8sServerCertificate.Secure(
-                                              _,
-                                              disableHostnameVerification
-                                            ) =>
-                                          disableHostnameVerification
-                                      }
-        client                     <-
-          ZManaged
-            .acquireReleaseAttemptWith(
+  val k8sSttpClient: ZLayer[K8sClusterConfig with Scope with Any, Throwable, SttpClient] = ZLayer {
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/asynchttpclient/package.scala:18]: The layer should not require `Scope` as an input - rather you should define the layer with `ZLayer.scoped { ... }` which provides a `Scope` for the layer's lifetime. That's equivalent of the previous code of constructing the layer from a `ZManaged`.

Also you should remove `with Any` as it does not add anything.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/asynchttpclient/package.scala [hunk 9]
```diff
@@ -52,20 +47,28 @@ package object asynchttpclient {
...
+      case K8sServerCertificate.Secure(_, disableHostnameVerification) =>
+        disableHostnameVerification
+    }
 
   /** Layer producing a [[K8sCluster]] and an [[SttpClient]] module that can be directly used to
     * initialize specific Kubernetes client modules, using the [[defaultConfigChain]].
     */
-  val k8sDefault: ZLayer[Any with System, Throwable, K8sCluster with SttpClient] =
-    (System.any) >+> defaultConfigChain >>> (k8sCluster ++ k8sSttpClient)
+  val k8sDefault: ZLayer[Any with Scope, Throwable, K8sCluster with SttpClient] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/asynchttpclient/package.scala:72]: Same things as for `k8sSttpClient`

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/httpclient/package.scala [hunk 10]
```diff
@@ -14,47 +14,54 @@ package object httpclient {
...
+                                         }
+      } yield client
+    }
 
   /** Layer producing a [[K8sCluster]] and an [[SttpClient]] module that can be directly used to
     * initialize specific Kubernetes client modules, using the [[defaultConfigChain]].
     */
-  val k8sDefault: ZLayer[Any with System, Throwable, K8sCluster with SttpClient] =
-    (System.any) >+> defaultConfigChain >>> (k8sCluster ++ k8sSttpClient)
+  val k8sDefault: ZLayer[Any with Scope, Throwable, K8sCluster with SttpClient] =
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/httpclient/package.scala:58]: Do `ZLayer.scoped` instead, fix type signature

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala [hunk 11]
```diff
@@ -256,30 +258,34 @@ package object config extends Descriptors {
...
     context: Option[String] = None,
     debug: Boolean = false,
     disableHostnameVerification: Boolean = false
-  ): ZLayer[Any with System with Path, Throwable, K8sClusterConfig] =
-    (for {
-      path   <- ZIO.service[Path]
-      config <- fromKubeconfigFile(path, context, debug, disableHostnameVerification)
-    } yield config).toLayer
+  ): ZLayer[Any with Path, Throwable, K8sClusterConfig] =
+    ZLayer.fromZIO {
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala:283]: You can just write `ZLayer {`

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala [hunk 12]
```diff
@@ -563,19 +569,19 @@ package object config extends Descriptors {
...
-  private def loadKeyString(source: KeySource): ZManaged[Any, Throwable, String] =
+  private def loadKeyString(source: KeySource): ZIO[Scope, Throwable, String] =
     source match {
       case KeySource.FromFile(path)     =>
-        ZManaged
+        ZIO
           .fromAutoCloseable(Task.attempt(new FileInputStream(path.toFile)))
           .flatMap { stream =>
-            ZManaged.fromZIO(Task(new String(stream.readAllBytes(), StandardCharsets.US_ASCII)))
+            ZIO.scoped(Task.attempt(new String(stream.readAllBytes(), StandardCharsets.US_ASCII)))
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala:578]: No need for this `ZIO.scoped` here

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala [hunk 13]
```diff
@@ -563,19 +569,19 @@ package object config extends Descriptors {
...
           }
       case KeySource.FromBase64(base64) =>
-        ZManaged.fromZIO(
-          Task(new String(Base64.getDecoder.decode(base64), StandardCharsets.US_ASCII))
+        ZIO.scoped(
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala:581]: No need for this `ZIO.scoped` here

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala [hunk 14]
```diff
@@ -563,19 +569,19 @@ package object config extends Descriptors {
...
+          Task.attempt(new String(Base64.getDecoder.decode(base64), StandardCharsets.US_ASCII))
         )
       case KeySource.FromString(value)  =>
-        ZManaged.succeed(value)
+        ZIO.scoped(ZIO.succeed(value))
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/package.scala:585]: No need for `ZIO.scoped` here

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/impl/ResourceClient.scala [hunk 15]
```diff
@@ -66,7 +66,7 @@ final class ResourceClient[
       }.map { initialResponse =>
         val rest = ZStream.fromPull {
           for {
-            nextContinueToken <- Ref.make(initialResponse.metadata.flatMap(_.continue)).toManaged
+            nextContinueToken <- ZIO.scoped(Ref.make(initialResponse.metadata.flatMap(_.continue)))
```
Comment (vigoo) [inline zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/impl/ResourceClient.scala:69]: No need for `ZIO.scoped` here

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaderForLifeLock.scala [hunk 16]
```diff
@@ -47,21 +47,21 @@ abstract class LeaderForLifeLock[T: K8sObject](
   def acquireLock(
     namespace: K8sNamespace,
     self: Pod
-  ): ZManaged[Clock, LeaderElectionFailure[Nothing], Unit] =
+  ): ZIO[Scope, LeaderElectionFailure[Nothing], Unit] =
     for {
-      alreadyOwned <- checkIfAlreadyOwned(namespace, self).toManaged
+      alreadyOwned <- ZIO.scoped(checkIfAlreadyOwned(namespace, self))
```
Comment (vigoo) [inline zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaderForLifeLock.scala:52]: No need for `ZIO.scoped`

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaseLock.scala [hunk 17]
```diff
@@ -29,24 +28,21 @@ class LeaseLock(
   override def acquireLock(
     namespace: K8sNamespace,
     pod: Pod
-  ): ZManaged[Clock, leader.LeaderElectionFailure[Nothing], Unit] =
+  ): ZIO[Scope, leader.LeaderElectionFailure[Nothing], Unit] =
     (for {
-      store <- Ref.makeManaged(Option.empty[VersionedRecord])
-      name  <- pod.getName.mapError(KubernetesError.apply).toManaged
+      store <- ZIO.scoped(Ref.make(Option.empty[VersionedRecord]))
```
Comment (vigoo) [inline zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaseLock.scala:33]: No need for `scoped` here either

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala [hunk 18]
```diff
@@ -19,28 +19,28 @@ package object leader {
...
         * the inner effect will only run at once in the Kubernetes cluster.
         *
-        * If you want to manage the lock as a ZManaged use [[lease]]
+        * If you want to manage the lock as Scoped use [[lease]]
         *
         * @param f
         *   Inner effect to protect
         */
-      def runAsLeader[R, E, A](f: ZIO[R, E, A]): ZIO[R with Clock, E, Option[A]] =
+      def runAsLeader[R, E, A](f: ZIO[R, E, A]): ZIO[R with Scope, E, Option[A]] =
```
Comment (vigoo) [inline zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala:27]: Same thing here about extending leaking out the scope. This should be a `ZIO` not requiring a `Scope`, and a `ZIO.scoped` used around the `lease` inside.

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala [hunk 19]
```diff
@@ -19,28 +19,28 @@ package object leader {
...
 
       /** Creates a managed lock implementing the leader election algorithm
         */
-      def lease: ZManaged[Clock, LeaderElectionFailure[Nothing], Unit]
+      def lease: ZIO[Scope, LeaderElectionFailure[Nothing], Unit]
     }
 
     class Live(contextInfo: ContextInfo.Service, lock: LeaderLock) extends Service {
-      override def lease: ZManaged[Clock, LeaderElectionFailure[Nothing], Unit] =
+      override def lease: ZIO[Scope, LeaderElectionFailure[Nothing], Unit] =
```
Comment (vigoo) [inline zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala:40]: Same things apply to this as mentioned earlier, about the use of `ZIO.scoped`

**Review by vigoo — CHANGES_REQUESTED** [review]
Added some comments related to some confusion around how to migrate the `ZManaged` code to `Scope`s. 

Thanks for your work!

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala (hunk 2 above)
Comment (gavares) [reply zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala:16]: Agreed, I think find/replace got the better of me.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/impl/ResourceClient.scala (hunk 15 above)
Comment (gavares) [reply zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/impl/ResourceClient.scala:69]: Isn't `ZIO.scoped` the replacement for `.toManaged`?

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaderForLifeLock.scala (hunk 16 above)
Comment (gavares) [reply zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaderForLifeLock.scala:52]: same comment as above

File: examples/logs-example/src/main/scala/com/coralogix/zio/k8s/examples/logs/LogsExample.scala (hunk 1 above)
Comment (gavares) [reply examples/logs-example/src/main/scala/com/coralogix/zio/k8s/examples/logs/LogsExample.scala]: Done

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala (hunk 3 above)
Comment (gavares) [reply zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala]: done.

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaderForLifeLock.scala (hunk 16 above)
Comment (gavares) [reply zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaderForLifeLock.scala]: done

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala (hunk 18 above)
Comment (gavares) [reply zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala]: done.

**Comment by gavares** [comment]
All tests now passing

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala (hunk 19 above)
Comment (gavares) [reply zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala]: done.

File: zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaseLock.scala (hunk 17 above)
Comment (gavares) [reply zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaseLock.scala]: done.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/impl/ResourceClient.scala (hunk 15 above)
Comment (gavares) [reply zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/impl/ResourceClient.scala]: updated.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala (hunk 4 above)
Comment (gavares) [reply zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala]: done.

File: zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala (hunk 2 above)
Comment (gavares) [reply zio-k8s-client/src/main/scala/com/coralogix/zio/k8s/client/config/KeyManagers.scala]: Done.

**Review by vigoo — APPROVED** [review]

**Pull request merged**